File: src/util/progress.ts
// Single-line progress renderer.
import { c, sym } from "./colors.js";

const isTTY = !!process.stdout.isTTY;
const FRAMES = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];

export interface Step {
  label: string;
  run: () => Promise<void> | void;
}

export class Progress {
  private idx = 0;
  private total = 0;
  private current = "";
  private timer: ReturnType<typeof setInterval> | null = null;
  private frame = 0;
  private startedAt = 0;
  private realWrite = process.stdout.write.bind(process.stdout);

  constructor(private title: string) {}

  /** Start the renderer; reserves a line. */
  start(total: number): void {
    this.total = total;
    this.idx = 0;
    this.startedAt = Date.now();
    if (this.title) console.log("\n  " + c.bold(c.cyan(this.title)));
    if (isTTY) {
      this.timer = setInterval(() => {
        this.frame = (this.frame + 1) % FRAMES.length;
        this.repaint();
      }, 80);
    }
  }

  /** Begin a sub-step; renders spinner + label until complete() or fail(). */
  begin(label: string): void {
    this.current = label;
    if (isTTY) this.repaint();
  }

  /** Mark current step as completed with a green check + optional note. */
  complete(note?: string): void {
    this.idx++;
    this.clearLine();
    const pct = this.total > 0 ? Math.round((this.idx / this.total) * 100) : 100;
    const noteStr = note ? c.gray(` ${note}`) : "";
    console.log(
      `  ${c.green(sym.check)} ${this.current.padEnd(22)} ${c.gray(`[${this.pctBar(pct)}] ${pct}%`)}${noteStr}`,
    );
  }

  /** Mark current step as failed. */
  fail(reason: string): void {
    this.idx++;
    this.clearLine();
    console.log(`  ${c.red(sym.cross)} ${this.current.padEnd(22)} ${c.red(reason)}`);
  }

  /** Mark current step as skipped (counts toward progress but neutral). */
  skip(note: string): void {
    this.idx++;
    this.clearLine();
    const pct = this.total > 0 ? Math.round((this.idx / this.total) * 100) : 100;
    console.log(
      `  ${c.gray(sym.bullet)} ${this.current.padEnd(22)} ${c.gray(`[${this.pctBar(pct)}] ${pct}%  ${note}`)}`,
    );
  }

  /** Finish the renderer. */
  done(summary?: string): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    this.clearLine();
    if (summary) console.log("  " + c.gray(summary));
  }

  // --- helpers ---
  private repaint(): void {
    if (!isTTY || !this.current) return;
    const f = c.cyan(FRAMES[this.frame]);
    const pct = this.total > 0 ? Math.round((this.idx / this.total) * 100) : 0;
    const line = `  ${f} ${this.current.padEnd(22)} ${c.gray(`[${this.pctBar(pct)}] ${pct}%`)}`;
    this.realWrite("\r\x1b[2K" + line);
  }

  private clearLine(): void {
    if (isTTY) this.realWrite("\r\x1b[2K");
  }

  private pctBar(pct: number): string {
    const width = 16;
    const filled = Math.round((pct / 100) * width);
    return c.green("█".repeat(filled)) + c.gray("░".repeat(width - filled));
  }
}

/** Wrap an async fn so any chatty logs it makes are silenced for the duration. */
export async function withSilencedLogs<T>(fn: () => Promise<T> | T): Promise<T> {
  const realStdout = process.stdout.write.bind(process.stdout);
  const realStderr = process.stderr.write.bind(process.stderr);
  const buf: string[] = [];
  process.stdout.write = ((chunk: string | Buffer) => {
    buf.push(typeof chunk === "string" ? chunk : chunk.toString("utf8"));
    return true;
  }) as typeof process.stdout.write;
  process.stderr.write = process.stdout.write;
  try {
    return await fn();
  } finally {
    process.stdout.write = realStdout;
    process.stderr.write = realStderr;
  }
}

File: internal/util/progress.go
package util

import (
	"fmt"
	"os"
	"strings"
	"sync"
	"time"
)

var frames = []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"}

// Progress renders a single-line spinner + step list.
type Progress struct {
	title   string
	idx     int
	total   int
	current string
	frame   int
	mu      sync.Mutex
	stop    chan struct{}
	tty     bool
}

func NewProgress(title string) *Progress {
	return &Progress{title: title, tty: stdoutIsTTY()}
}

func (p *Progress) Start(total int) {
	p.total = total
	p.idx = 0
	if p.title != "" {
		fmt.Println("\n  " + C.Bold(C.Cyan(p.title)))
	}
	if p.tty {
		p.stop = make(chan struct{})
		go p.spin()
	}
}

func (p *Progress) spin() {
	t := time.NewTicker(80 * time.Millisecond)
	defer t.Stop()
	for {
		select {
		case <-p.stop:
			return
		case <-t.C:
			p.mu.Lock()
			p.frame = (p.frame + 1) % len(frames)
			p.repaint()
			p.mu.Unlock()
		}
	}
}

func (p *Progress) Begin(label string) {
	p.mu.Lock()
	p.current = label
	if p.tty {
		p.repaint()
	}
	p.mu.Unlock()
}

func padEnd(s string, n int) string {
	if len(s) >= n {
		return s
	}
	return s + strings.Repeat(" ", n-len(s))
}

func (p *Progress) pctBar(pct int) string {
	width := 16
	filled := pct * width / 100
	if filled > width {
		filled = width
	}
	return C.Green(strings.Repeat("█", filled)) + C.Gray(strings.Repeat("░", width-filled))
}

func (p *Progress) Complete(note string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.idx++
	p.clearLine()
	pct := 100
	if p.total > 0 {
		pct = p.idx * 100 / p.total
	}
	noteStr := ""
	if note != "" {
		noteStr = C.Gray(" " + note)
	}
	fmt.Printf("  %s %s %s%s\n", C.Green(Sym.Check), padEnd(p.current, 22),
		C.Gray(fmt.Sprintf("[%s] %d%%", p.pctBar(pct), pct)), noteStr)
}

func (p *Progress) Fail(reason string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.idx++
	p.clearLine()
	fmt.Printf("  %s %s %s\n", C.Red(Sym.Cross), padEnd(p.current, 22), C.Red(reason))
}

func (p *Progress) Skip(note string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.idx++
	p.clearLine()
	pct := 100
	if p.total > 0 {
		pct = p.idx * 100 / p.total
	}
	fmt.Printf("  %s %s %s\n", C.Gray(Sym.Bullet), padEnd(p.current, 22),
		C.Gray(fmt.Sprintf("[%s] %d%%  %s", p.pctBar(pct), pct, note)))
}

func (p *Progress) Done(summary string) {
	if p.stop != nil {
		close(p.stop)
		p.stop = nil
	}
	p.mu.Lock()
	p.clearLine()
	p.mu.Unlock()
	if summary != "" {
		fmt.Println("  " + C.Gray(summary))
	}
}

func (p *Progress) repaint() {
	if !p.tty || p.current == "" {
		return
	}
	pct := 0
	if p.total > 0 {
		pct = p.idx * 100 / p.total
	}
	line := fmt.Sprintf("  %s %s %s", C.Cyan(frames[p.frame]), padEnd(p.current, 22),
		C.Gray(fmt.Sprintf("[%s] %d%%", p.pctBar(pct), pct)))
	fmt.Fprint(os.Stdout, "\r\x1b[2K"+line)
}

func (p *Progress) clearLine() {
	if p.tty {
		fmt.Fprint(os.Stdout, "\r\x1b[2K")
	}
}

// WithSilencedLogs redirects stdout/stderr to a buffer while fn runs.
func WithSilencedLogs(fn func() error) error {
	realOut, realErr := os.Stdout, os.Stderr
	r, w, _ := os.Pipe()
	os.Stdout, os.Stderr = w, w
	done := make(chan struct{})
	go func() {
		buf := make([]byte, 4096)
		for {
			if _, err := r.Read(buf); err != nil {
				break
			}
		}
		close(done)
	}()
	defer func() {
		w.Close()
		os.Stdout, os.Stderr = realOut, realErr
		<-done
		r.Close()
	}()
	return fn()
}
